File: Cpp-C/Ema/Src/Rdm/Impl/EnumType.cpp
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2017,2019-2020 LSEG. All rights reserved.
 *|-----------------------------------------------------------------------------
 */

#include "EnumType.h"
#include "EnumTypeImpl.h"
#include "ExceptionTranslator.h"

#include <new>

using namespace refinitiv::ema::rdm;

EnumType::EnumType()
{
	try
	{
		_pImpl = new EnumTypeImpl();
	}
	catch (std::bad_alloc&)
	{
		throwMeeException("Failed to allocate memory in EnumType::EnumType()");
	}
}

EnumType& EnumType::operator=(const EnumType& other)
{
	this->_pImpl->rsslEnumType(other._pImpl->rsslEnumType());
	return *this;
}

EnumType::~EnumType()
{
	if ( _pImpl )
	{
		delete _pImpl;
		_pImpl = 0;
	}
}

refinitiv::ema::access::UInt16 EnumType::getValue() const
{
	return _pImpl->getValue();
}

const refinitiv::ema::access::EmaString& EnumType::getDisplay() const
{
	return _pImpl->getDisplay();
}

const refinitiv::ema::access::EmaString& EnumType::getMeaning() const
{
	return _pImpl->getMeaning();
}

const refinitiv::ema::access::EmaString& EnumType::toString() const
{
	return _pImpl->toString();
}

EnumType::operator const char* () const
{
	return _pImpl->toString().c_str();
}

File: Java/Eta/Core/src/main/java/com/refinitiv/eta/codec/EnumType.java
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2019-2022 LSEG. All rights reserved.     
 *|-----------------------------------------------------------------------------
 */

package com.refinitiv.eta.codec;

import com.refinitiv.eta.codec.Buffer;

/** 
 * A single defined enumerated value.
 */
public interface EnumType 
{
    /**
     * The actual value representing the type.
     * 
     * @return the value
     */
    public int value();

    /**
     * A brief string representation describing what the type means (For example,
     * this may be an abbreviation of a currency to be displayed to a user).
     *
     * Note: toString() method called on the display buffer
     * will return String object obtained from the underlying bytes using ISO-8859-1 encoding regardless of the
     * encoding set for JVM in order to avoid data corruption in case the bytes represent an RMTES string incompatible
     * with, e.g., UTF-8 encoding
     *
     * 
     * @return the display
     */
    public Buffer display();

    /**
     * A more elaborate description of what the value means. This information is
     * typically optional and not displayed.
     * 
     * @return the meaning
     */
    public Buffer meaning();
}